dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/report/TrackerValidationReport.java

```diff
@@ -28,85 +28,146 @@
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
 
-import lombok.AllArgsConstructor;
-import lombok.Builder;
-import lombok.Data;
-import lombok.NoArgsConstructor;
+import lombok.EqualsAndHashCode;
+import lombok.ToString;
+
+import org.hisp.dhis.tracker.TrackerType;
+import org.hisp.dhis.tracker.domain.TrackerDto;
 
 import com.fasterxml.jackson.annotation.JsonIgnore;
 import com.fasterxml.jackson.annotation.JsonProperty;
 
 /**
  * @author Morten Olav Hansen <mortenoh@gmail.com>
  */
-@Data
-@Builder
-@NoArgsConstructor
-@AllArgsConstructor
+@ToString
+@EqualsAndHashCode
 public class TrackerValidationReport
 {
     @JsonProperty
-    @Builder.Default
-    private List<TrackerErrorReport> errorReports = new ArrayList<>();
+    private final List<TrackerErrorReport> errorReports;
 
     @JsonProperty
-    @Builder.Default
-    private List<TrackerWarningReport> warningReports = new ArrayList<>();
+    private final List<TrackerWarningReport> warningReports;
+
+    @JsonIgnore
+    private final List<TrackerValidationHookTimerReport> performanceReport;
 
+    /*
+     * Keeps track of all the invalid Tracker objects (i.e. objects with at
+     * least one TrackerErrorReport in the TrackerValidationReport) encountered
+     * during the validation process.
+     */
     @JsonIgnore
-    @Builder.Default
-    private List<TrackerValidationHookTimerReport> performanceReport = new ArrayList<>();
+    private final Map<TrackerType, List<String>> invalidDTOs;
+
+    public TrackerValidationReport()
+    {
+        this.errorReports = new ArrayList<>();
+        this.warningReports = new ArrayList<>();
+        this.performanceReport = new ArrayList<>();
+        this.invalidDTOs = new HashMap<>();
+    }
 
     // -----------------------------------------------------------------------------------
     // Utility Methods
     // -----------------------------------------------------------------------------------
 
-    public void add( TrackerValidationReport validationReport )
+    public void addValidationReport( TrackerValidationReport report )
+    {
+        addErrors( report.getErrorReports() );
+        addWarnings( report.getWarningReports() );
+        addPerfReports( report.getPerformanceReport() );
+    }
+
+    public List<TrackerErrorReport> getErrorReports()
+    {
+        return Collections.unmodifiableList( errorReports );
+    }
+
+    public List<TrackerWarningReport> getWarningReports()
+    {
+        return Collections.unmodifiableList( warningReports );
+    }
+
+    public List<TrackerValidationHookTimerReport> getPerformanceReport()
     {
-        add( validationReport.getErrorReports() );
-        addWarnings( validationReport.getWarningReports() );
-        addPerfReports( validationReport.getPerformanceReport() );
+        return Collections.unmodifiableList( performanceReport );
+    }
+
+    public void addError( TrackerErrorReport error )
```

Comment:
I think it makes sense to have this adders return `this` so one can use them chaining.
This helps turn typical multi-liners like

```java
SomeReport report = new SomeReport();
report.addError(error);
return report;
``` 
into 
```java
return new SomeReport().addError(error);
```

Reply:
Will do that in a follow up! :+1: